src/renderer/components/swap/Swap.tsx

```diff
@@ -734,6 +749,57 @@ export const Swap = ({
     )
   }, [canSwitchAssets, setPendingSwitchAssets, assetsToSwap, onChangePath])
 
+  const [isApprovedRD] = useObservableState<RD.RemoteData<ApiError, boolean>>(
+    () =>
+      FP.pipe(
+        sourcePoolRouter,
+        O.map((router) => {
+          if (isERC20Asset(sourceAssetProp)) {
+            return isApprovedERC20Token$({
+              spender: router,
+              sender: getEthTokenAddress(ERC20Assets.filter((asset) => eqAsset.equals(asset, sourceAssetProp))[0])
+            })
+          } else {
+            return Rx.of(RD.success(true))
+          }
+        }),
+        O.getOrElse<LiveData<ApiError, boolean>>(() => Rx.of(RD.success(true)))
+      ),
+    RD.initial
+  )
+
+  const [isApproved, setApproved] = useState(true)
+
+  useEffect(() => {
+    if (RD.isSuccess(isApprovedRD)) {
+      setApproved(isApprovedRD.value)
+    } else {
+      setApproved(true)
+    }
+  }, [chainFeesRD, isApprovedRD])
+
+  const [isApproveLoading, setApproveLoading] = useState(false)
+  const [approveError, setApproveError] = useState('')
+
+  const onApprove = () => {
+    FP.pipe(
+      sourcePoolRouter,
+      O.map((router) =>
+        approveERC20Token$({
+          spender: router,
+          sender: getEthTokenAddress(ERC20Assets.filter((asset) => eqAsset.equals(asset, sourceAssetProp))[0])
+        }).subscribe((val) => {
+          setApproveLoading(RD.isPending(val))
+          if (RD.isFailure(val)) {
+            setApproveError(val.error.msg)
+          } else if (RD.isSuccess(val)) {
+            setApproved(true)
+          }
+        })
```

Comment:
If we subscribe to a stream, we do need a way unsubscribe if the component is going to be unmounted at latest.